### PR TITLE
Added option to disable webhooks

### PR DESCRIPTION
**Pull request by pboguslawski**
This mod introduces DISABLE_WEBHOOKS parameter in [security] section
of app.ini (by default set to false). If set to true it disables web
hooks feature. Any existing undelivered web hook tasks will be cancelled.
Any existing web hook definitions will be left untouched in db but
its delivery tasks will be ignored.

### CODE REVIEW

**Review by techknowlogick — DISMISSED** [review]
`Webhook` is one word (per https://en.wikipedia.org/wiki/Webhook) could you update this PR to changes references to have it be one word instead of two?

File: custom/conf/app.example.ini [hunk 1]
```diff
@@ -541,6 +541,8 @@ IMPORT_LOCAL_PATHS = false
 ; It also enables them to access other resources available to the user on the operating system that is running the Gitea instance and perform arbitrary actions in the name of the Gitea OS user.
 ; WARNING: This maybe harmful to you website or your operating system.
 DISABLE_GIT_HOOKS = true
+; Set to false to disable web hooks feature.
+DISABLE_WEB_HOOKS = false
```
Comment (zeripath) [inline custom/conf/app.example.ini]:
```suggestion
DISABLE_WEBHOOKS = false
```

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md [hunk 2]
```diff
@@ -400,6 +400,7 @@ relation to port exhaustion.
    It also enables them to access other resources available to the user on the operating system that is running the
    Gitea instance and perform arbitrary actions in the name of the Gitea OS user.
    This maybe harmful to you website or your operating system.
+- `DISABLE_WEB_HOOKS`: **false**: Set to `true` to disable web hooks feature.
```
Comment (zeripath) [inline docs/content/doc/advanced/config-cheat-sheet.en-us.md]:
```suggestion
- `DISABLE_WEBHOOKS`: **false**: Set to `true` to disable web hooks feature.
```

File: modules/setting/setting.go [hunk 3]
```diff
@@ -152,6 +152,7 @@ var (
 	MinPasswordLength                  int
 	ImportLocalPaths                   bool
 	DisableGitHooks                    bool
+	DisableWebHooks                    bool
```
Comment (zeripath) [inline modules/setting/setting.go]:
```suggestion
	DisableWebhooks                    bool
```

File: modules/setting/setting.go [hunk 4]
```diff
@@ -770,6 +771,7 @@ func NewContext() {
 	MinPasswordLength = sec.Key("MIN_PASSWORD_LENGTH").MustInt(6)
 	ImportLocalPaths = sec.Key("IMPORT_LOCAL_PATHS").MustBool(false)
 	DisableGitHooks = sec.Key("DISABLE_GIT_HOOKS").MustBool(true)
+	DisableWebHooks = sec.Key("DISABLE_WEB_HOOKS").MustBool(false)
```
Comment (zeripath) [inline modules/setting/setting.go]:
```suggestion
	DisableWebhooks = sec.Key("DISABLE_WEBHOOKS").MustBool(false)
```

File: modules/templates/helper.go [hunk 5]
```diff
@@ -225,6 +225,9 @@ func NewFuncMap() []template.FuncMap {
 		"DisableGitHooks": func() bool {
 			return setting.DisableGitHooks
 		},
+		"DisableWebHooks": func() bool {
+			return setting.DisableWebHooks
+		},
```
Comment (zeripath) [inline modules/templates/helper.go]:
```suggestion
		"DisableWebhooks": func() bool {
			return setting.DisableWebhooks
		},
```

File: modules/webhook/deliver.go [hunk 6]
```diff
@@ -141,6 +141,10 @@ func Deliver(t *models.HookTask) error {
 		}
 	}()
 
+	if setting.DisableWebHooks {
+		return fmt.Errorf("Sending webhook skipped (web hooks disabled): [%d]", t.ID)
+	}
```
Comment (zeripath) [inline modules/webhook/deliver.go]:
```suggestion
	if setting.DisableWebhooks {
		return fmt.Errorf("Sending webhook skipped (webhooks disabled): [%d]", t.ID)
	}
```

File: modules/webhook/webhook.go [hunk 7]
```diff
@@ -68,6 +68,11 @@ func checkBranch(w *models.Webhook, branch string) bool {
 }
 
 func prepareWebhook(w *models.Webhook, repo *models.Repository, event models.HookEventType, p api.Payloader) error {
+	// Skip sending if web hooks are disabled.
+	if setting.DisableWebHooks {
```
Comment (zeripath) [inline modules/webhook/webhook.go]:
```suggestion
	// Skip sending if webhooks are disabled.
	if setting.DisableWebhooks {
```

File: routers/api/v1/api.go [hunk 8]
```diff
@@ -374,6 +374,16 @@ func reqGitHook() macaron.Handler {
...
 
+// reqWebHooksEnabled requires web hooks to be enabled by admin.
+func reqWebHooksEnabled() macaron.Handler {
+	return func(ctx *context.APIContext) {
+		if setting.DisableWebHooks {
+			ctx.Error(http.StatusForbidden, "", "web hooks disabled by administrator")
+			return
+		}
+	}
+}
```
Comment (zeripath) [inline routers/api/v1/api.go]:
```suggestion
// reqWebhooksEnabled requires webhooks to be enabled by admin.
func reqWebhooksEnabled() macaron.Handler {
	return func(ctx *context.APIContext) {
		if setting.DisableWebhooks {
			ctx.Error(http.StatusForbidden, "", "webhooks disabled by administrator")
			return
		}
	}
}
```

File: routers/api/v1/api.go [hunk 9]
```diff
@@ -655,15 +673,7 @@ func RegisterRoutes(m *macaron.Macaron) {
...
-					m.Group("/git", func() {
-						m.Combo("").Get(repo.ListGitHooks)
-						m.Group("/:id", func() {
-							m.Combo("").Get(repo.GetGitHook).
-								Patch(bind(api.EditGitHookOption{}), repo.EditGitHook).
-								Delete(repo.DeleteGitHook)
-						})
-					}, reqGitHook(), context.ReferencesGitRepo(true))
-				}, reqToken(), reqAdmin())
+				}, reqToken(), reqAdmin(), reqWebHooksEnabled())
```
Comment (zeripath) [inline routers/api/v1/api.go]:
```suggestion
				}, reqToken(), reqAdmin(), reqWebhooksEnabled())
```

File: routers/api/v1/api.go [hunk 10]
```diff
@@ -914,7 +924,7 @@ func RegisterRoutes(m *macaron.Macaron) {
 				m.Combo("/:id").Get(org.GetHook).
 					Patch(bind(api.EditHookOption{}), org.EditHook).
 					Delete(org.DeleteHook)
-			}, reqToken(), reqOrgOwnership())
+			}, reqToken(), reqOrgOwnership(), reqWebHooksEnabled())
```
Comment (zeripath) [inline routers/api/v1/api.go]:
```suggestion
			}, reqToken(), reqOrgOwnership(), reqWebhooksEnabled())
```

File: routers/routes/routes.go [hunk 11]
```diff
@@ -331,6 +331,14 @@ func RegisterRoutes(m *macaron.Macaron) {
 		}
 	}
 
+	// webHooksEnabled requires web hooks to be enabled by admin.
+	webHooksEnabled := func(ctx *context.Context) {
+		if setting.DisableWebHooks {
+			ctx.Error(403)
+			return
+		}
+	}
```
Comment (zeripath) [inline routers/routes/routes.go]:
```suggestion
	// webhooksEnabled requires webhooks to be enabled by admin.
	webhooksEnabled := func(ctx *context.Context) {
		if setting.DisableWebhooks {
			ctx.Error(403)
			return
		}
	}
```

File: routers/routes/routes.go [hunk 12]
```diff
@@ -550,7 +558,7 @@ func RegisterRoutes(m *macaron.Macaron) {
 			m.Post("/matrix/:id", bindIgnErr(auth.NewMatrixHookForm{}), repo.MatrixHooksEditPost)
 			m.Post("/msteams/:id", bindIgnErr(auth.NewMSTeamsHookForm{}), repo.MSTeamsHooksEditPost)
 			m.Post("/feishu/:id", bindIgnErr(auth.NewFeishuHookForm{}), repo.FeishuHooksEditPost)
-		})
+		}, webHooksEnabled)
```
Comment (zeripath) [inline routers/routes/routes.go]:
```suggestion
		}, webhooksEnabled)
```

File: routers/routes/routes.go [hunk 13]
```diff
@@ -655,7 +663,7 @@ func RegisterRoutes(m *macaron.Macaron) {
 					m.Post("/matrix/:id", bindIgnErr(auth.NewMatrixHookForm{}), repo.MatrixHooksEditPost)
 					m.Post("/msteams/:id", bindIgnErr(auth.NewMSTeamsHookForm{}), repo.MSTeamsHooksEditPost)
 					m.Post("/feishu/:id", bindIgnErr(auth.NewFeishuHookForm{}), repo.FeishuHooksEditPost)
-				})
+				}, webHooksEnabled)
```
Comment (zeripath) [inline routers/routes/routes.go]:
```suggestion
				}, webhooksEnabled)
```

File: routers/routes/routes.go [hunk 14]
```diff
@@ -732,13 +746,7 @@ func RegisterRoutes(m *macaron.Macaron) {
...
 				m.Post("/msteams/:id", bindIgnErr(auth.NewMSTeamsHookForm{}), repo.MSTeamsHooksEditPost)
 				m.Post("/feishu/:id", bindIgnErr(auth.NewFeishuHookForm{}), repo.FeishuHooksEditPost)
-
-				m.Group("/git", func() {
-					m.Get("", repo.GitHooks)
-					m.Combo("/:name").Get(repo.GitHooksEdit).
-						Post(repo.GitHooksEditPost)
-				}, context.GitHookService())
-			})
+			}, webHooksEnabled)
```
Comment (zeripath) [inline routers/routes/routes.go]:
```suggestion
			}, webhooksEnabled)
```

File: templates/admin/navbar.tmpl [hunk 15]
```diff
@@ -11,12 +11,14 @@
 	<a class="{{if .PageIsAdminRepositories}}active{{end}} item" href="{{AppSubUrl}}/admin/repos">
 		{{.i18n.Tr "admin.repositories"}}
 	</a>
+	{{if not DisableWebHooks}}
```
Comment (zeripath) [inline templates/admin/navbar.tmpl]:
```suggestion
	{{if not DisableWebhooks}}
```

File: templates/org/settings/navbar.tmpl [hunk 16]
```diff
@@ -4,9 +4,11 @@
 		<a class="{{if .PageIsSettingsOptions}}active{{end}} item" href="{{.OrgLink}}/settings">
 			{{.i18n.Tr "org.settings.options"}}
 		</a>
+		{{if not DisableWebHooks}}
```
Comment (zeripath) [inline templates/org/settings/navbar.tmpl]:
```suggestion
		{{if not DisableWebhooks}}
```

File: templates/repo/settings/nav.tmpl [hunk 17]
```diff
@@ -5,7 +5,9 @@
 			<li {{if .PageIsSettingsOptions}}class="current"{{end}}><a href="{{.RepoLink}}/settings">{{.i18n.Tr "repo.settings.options"}}</a></li>
 			<li {{if .PageIsSettingsCollaboration}}class="current"{{end}}><a href="{{.RepoLink}}/settings/collaboration">{{.i18n.Tr "repo.settings.collaboration"}}</a></li>
 			<li {{if .PageIsSettingsBranches}}class="current"{{end}}><a href="{{.RepoLink}}/settings/branches">{{.i18n.Tr "repo.settings.branches"}}</a></li>
+			{{if not DisableWebHooks}}
```
Comment (zeripath) [inline templates/repo/settings/nav.tmpl]:
```suggestion
			{{if not DisableWebhooks}}
```

File: templates/repo/settings/navbar.tmpl [hunk 18]
```diff
@@ -10,9 +10,11 @@
 			{{.i18n.Tr "repo.settings.branches"}}
 		</a>
 	{{end}}
+	{{if not DisableWebHooks}}
```
Comment (zeripath) [inline templates/repo/settings/navbar.tmpl]:
```suggestion
	{{if not DisableWebhooks}}
```

**Comment by pboguslawski** [comment]
Spelling fixed.

File: custom/conf/app.example.ini [hunk 19]
```diff
@@ -541,6 +541,8 @@ IMPORT_LOCAL_PATHS = false
 ; It also enables them to access other resources available to the user on the operating system that is running the Gitea instance and perform arbitrary actions in the name of the Gitea OS user.
 ; WARNING: This maybe harmful to you website or your operating system.
 DISABLE_GIT_HOOKS = true
+; Set to false to disable webhooks feature.
```
Comment (6543) [inline custom/conf/app.example.ini]:
```suggestion
; Set to true to disable webhooks feature.
```

Comment (pboguslawski) [reply custom/conf/app.example.ini]: Parameter description fixed.

**Review by 6543 — APPROVED** [review]

**Comment by lafriks** [comment]
please resolve conflicts

**Comment by pboguslawski** [comment]
Master merged to this PR.

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
@pboguslawski can you update again?

**Comment by pboguslawski** [comment]
Master merged to this PR.

**Comment by 6543** [comment]
@pboguslawski can you merge master into this and apply latest changes (look at 6433ba0ec3dfde67f45267aa12bd713c4a44c740 for more infos)

**Comment by 6543** [comment]
@pboguslawski hope you dont mind if you could `merge master into it`/`resolve conflicts`

**Comment by pboguslawski** [comment]
Master merged to this PR.

**Review by khmarbaise — APPROVED** [review]

**Comment by 6543** [comment]
@pboguslawski need one update again :D

**Comment by pboguslawski** [comment]
Just pressed update button on this page.

**Pull request merged**